src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -1343,25 +1375,39 @@ void uct_dc_mlx5_ep_pending_purge(uct_ep_h tl_ep, uct_pending_purge_callback_t c
     }
 }
 
-ucs_status_t uct_dc_mlx5_ep_check_fc(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_t *ep)
+ucs_status_t uct_dc_mlx5_ep_check_fc(uct_dc_mlx5_iface_t *iface,
+                                     uct_dc_mlx5_ep_t *ep)
 {
     ucs_status_t status;
 
-    if (iface->super.super.config.fc_enabled) {
-        UCT_RC_CHECK_FC_WND(&ep->fc, ep->super.stats);
-        if (ep->fc.fc_wnd == iface->super.super.config.fc_hard_thresh) {
-            status = uct_rc_fc_ctrl(&ep->super.super,
-                                    UCT_RC_EP_FLAG_FC_HARD_REQ,
-                                    NULL);
-            if (status != UCS_OK) {
-                return status;
-            }
-        }
-    } else {
+    if (!iface->super.super.config.fc_enabled) {
         /* Set fc_wnd to max, to send as much as possible without checks */
         ep->fc.fc_wnd = INT16_MAX;
+        return UCS_OK;
     }
-    return UCS_OK;
+
+    if (ucs_likely((ep->fc.fc_wnd > 0) &&
+                   (ep->fc.fc_wnd !=
+                            iface->super.super.config.fc_hard_thresh))) {
+        /* Do not send a grant request */
+        return UCS_OK;
+    }
+
+    status = uct_rc_fc_ctrl(&ep->super.super, UCT_RC_EP_FLAG_FC_HARD_REQ,
+                            NULL);
+    if ((status == UCS_OK) && (ep->fc.fc_wnd <= 0)) {
+        status = UCS_ERR_NO_RESOURCE;
+    }
+
+    if ((ep->dci != UCT_DC_MLX5_EP_NO_DCI) && 
+        !uct_dc_mlx5_iface_is_dci_rand(iface)) {
+        ucs_assertv(uct_dc_mlx5_iface_dci_has_outstanding(iface, ep->dci),
+                    "iface %p ep %p: dci leak detected: dci=%d fc_wnd=%d",
+                    iface, ep, ep->dci, ep->fc.fc_wnd);
+    }
+
+    return ((ep->fc.fc_wnd > 0) || (status != UCS_OK)) ?
+           status : UCS_ERR_NO_RESOURCE;
```

Comment:
why not return `status` here?

Reply:
we should return NO_RESOURCE always for `fc_wnd <= 0`

Reply:
if status is not ok (e.g. `UCS_ERR_NO_MEMORY`), you can return it instead. What is the caveat of that?
The pro is avoiding extra branch + more readable code